### PR TITLE
Bugfix: Scheduler monitor for manual optimization

### PR DESCRIPTION
**Pull request by maxoppelt**
## What does this PR do?

Fixes #6567

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/common/optimizers.rst [hunk 1]
```diff
@@ -230,6 +230,23 @@ If you want to call ``lr_scheduler.step()`` every ``n`` steps/epochs, do the fol
         if self.trainer.is_last_batch and (self.trainer.current_epoch + 1) % n == 0:
             sch.step()
 
+If you want to call optimizers that require a metric value after each epoch, consider doing the following:
```
Comment (rohitgr7) [inline docs/source/common/optimizers.rst:233]:
```suggestion
If you want to call schedulers that require a metric value after each epoch, consider doing the following:
```
you mean schedulers here?

File: docs/source/common/optimizers.rst [hunk 2]
```diff
@@ -230,6 +230,23 @@ If you want to call ``lr_scheduler.step()`` every ``n`` steps/epochs, do the fol
...
+
+.. testcode:: python
```
Comment (rohitgr7) [inline docs/source/common/optimizers.rst:235]:
```suggestion
.. testcode::
```

File: docs/source/common/optimizers.rst [hunk 3]
```diff
@@ -230,6 +230,23 @@ If you want to call ``lr_scheduler.step()`` every ``n`` steps/epochs, do the fol
...
+
+    def __init__(self):
+        super().__init__()
+        self.automatic_optimization = False
+
+    def training_epoch_end(self, outputs):
+        sch = self.lr_schedulers()
+
+        # If the selected scheduler is a ReduceLROnPlateau scheduler.
+        if isinstance(sch, th.optim.lr_scheduler.ReduceLROnPlateau):
```
Comment (rohitgr7) [inline docs/source/common/optimizers.rst:245]:
```suggestion
        if isinstance(sch, torch.optim.lr_scheduler.ReduceLROnPlateau):
```

File: pytorch_lightning/trainer/optimizers.py [hunk 4]
```diff
@@ -150,13 +150,16 @@ def configure_schedulers(
                 scheduler['reduce_on_plateau'] = isinstance(
                     scheduler['scheduler'], optim.lr_scheduler.ReduceLROnPlateau
                 )
-                if scheduler['reduce_on_plateau'] and scheduler.get('monitor', None) is None:
+                if scheduler['reduce_on_plateau'] and scheduler.get('monitor', None) is None \
+                        and not is_manual_optimization:
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/optimizers.py:154]:
```suggestion
                if (
                    scheduler['reduce_on_plateau'] and scheduler.get('monitor', None) is None
                    and not is_manual_optimization
                ):
```

File: pytorch_lightning/trainer/optimizers.py [hunk 5]
```diff
@@ -150,13 +150,16 @@ def configure_schedulers(
...
                     raise MisconfigurationException(
                         'The lr scheduler dict must include a monitor when a `ReduceLROnPlateau` scheduler is used.'
                         ' For example: {"optimizer": optimizer, "lr_scheduler":'
                         ' {"scheduler": scheduler, "monitor": "your_loss"}}'
                     )
                 lr_schedulers.append({**default_config, **scheduler})
+            elif isinstance(scheduler, optim.lr_scheduler.ReduceLROnPlateau) and is_manual_optimization:
+                lr_schedulers.append({**default_config, 'scheduler': scheduler})
             elif isinstance(scheduler, optim.lr_scheduler.ReduceLROnPlateau):
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/optimizers.py:163]: logic is correct but maybe combine both of them into one?

Comment (maxoppelt) [reply pytorch_lightning/trainer/optimizers.py]: Ok, resolved in https://github.com/PyTorchLightning/pytorch-lightning/pull/7643/commits/6a7e3009785851f220d706dc523118fb7046e2be.

However, also invalid keys are required in pytorch lightning, replacing invalid with default parameters.

File: docs/source/common/optimizers.rst [hunk 6]
```diff
@@ -230,6 +230,23 @@ If you want to call ``lr_scheduler.step()`` every ``n`` steps/epochs, do the fol
...
+    def __init__(self):
+        super().__init__()
+        self.automatic_optimization = False
+
+    def training_epoch_end(self, outputs):
+        sch = self.lr_schedulers()
+
+        # If the selected scheduler is a ReduceLROnPlateau scheduler.
+        if isinstance(sch, torch.optim.lr_scheduler.ReduceLROnPlateau):
+            loss = th.stack([x['loss'] for x in outputs]).mean()
```
Comment (carmocca) [inline docs/source/common/optimizers.rst:246]:
```suggestion
            loss = torch.stack([x['loss'] for x in outputs]).mean()
```

**Review by carmocca — CHANGES_REQUESTED** [review]
Can you add a test and update the CHANGELOG.md file?

File: docs/source/common/optimizers.rst (hunk 6 above)
Comment (maxoppelt) [reply docs/source/common/optimizers.rst:246]: Fixed package pytorch name.

File: tests/trainer/optimization/test_manual_optimization.py [hunk 7]
```diff
@@ -992,6 +992,65 @@ def configure_optimizers(self):
...
+                    'scheduler': torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer_2),
+                    'monitor': 'train_loss_2'
+                }
+            else:
+                scheduler_1 = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer_1)
+                scheduler_2 = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer_2)
+
+            return [optimizer_1, optimizer_2], [scheduler_1, scheduler_2]
+
+    for test_scheduler_as_dict in [True, False]:
```
Comment (SkafteNicki) [inline tests/trainer/optimization/test_manual_optimization.py:1040]: when `test_scheduler_as_dict=True` the test should probably have a `pytest.raises` because `monitor` belongs to the set of invalid when doing manual optimization

File: docs/source/common/optimizers.rst [hunk 8]
```diff
@@ -230,6 +230,65 @@ If you want to call ``lr_scheduler.step()`` every ``n`` steps/epochs, do the fol
...
+            loss = self.all_gather(loss).mean()
+            sch.step(loss)
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply docs/source/common/optimizers.rst]: That shouldn't have been part of this PR. Updated!

File: tests/trainer/optimization/test_manual_optimization.py (hunk 7 above)
Comment (kaushikb11) [reply tests/trainer/optimization/test_manual_optimization.py]: Test updated!

File: tests/trainer/optimization/test_manual_optimization.py [hunk 9]
```diff
@@ -957,6 +957,59 @@ def configure_optimizers(self):
...
+            return {"train_loss_1": torch.tensor([0.0]), "train_loss_2": torch.tensor([0.0])}
+
+        def training_epoch_end(self, outputs):
+            scheduler_1, scheduler_2 = self.lr_schedulers()
+
+            loss = torch.stack([x["train_loss_1"] for x in outputs]).mean()
+            scheduler_1.step(loss)
+
+            loss = torch.stack([x["train_loss_2"] for x in outputs]).mean()
+            scheduler_1.step(loss)
```
Comment (tchaton) [inline tests/trainer/optimization/test_manual_optimization.py:978]: Can we mock the schedulers and make sure they are call the right number of times ?

Comment (tchaton) [inline tests/trainer/optimization/test_manual_optimization.py:978]:
```suggestion
            scheduler_2.step(loss)
```

File: docs/source/common/optimizers.rst [hunk 10]
```diff
@@ -230,6 +230,23 @@ If you want to call ``lr_scheduler.step()`` every ``n`` steps/epochs, do the fol
...
+.. testcode::
+
+    def __init__(self):
+        super().__init__()
+        self.automatic_optimization = False
+
+    def training_epoch_end(self, outputs):
+        sch = self.lr_schedulers()
+
+        # If the selected scheduler is a ReduceLROnPlateau scheduler.
```
Comment (tchaton) [inline docs/source/common/optimizers.rst:244]: Might be better to rely on logging to perform syncing.

Comment (tchaton) [reply docs/source/common/optimizers.rst:244]: sch.step(self.trainer.callback_metrics["loss"])

File: tests/trainer/optimization/test_manual_optimization.py [hunk 11]
```diff
@@ -957,6 +957,59 @@ def configure_optimizers(self):
...
+        def training_epoch_end(self, outputs):
+            scheduler_1, scheduler_2 = self.lr_schedulers()
+
+            loss = torch.stack([x["train_loss_1"] for x in outputs]).mean()
+            scheduler_1.step(loss)
+
+            loss = torch.stack([x["train_loss_2"] for x in outputs]).mean()
+            scheduler_2.step(loss)
+
+        def configure_optimizers(self):
```
Comment (awaelchli) [inline tests/trainer/optimization/test_manual_optimization.py:980]: are two schedulers required here? we already have tests with muliple schedulers + manual right?

File: pytorch_lightning/trainer/optimizers.py [hunk 12]
```diff
@@ -119,21 +119,8 @@ def configure_schedulers(
         lr_schedulers = []
         default_config = _get_default_scheduler_config()
         for scheduler in schedulers:
-            if isinstance(scheduler, dict):
```
Comment (awaelchli) [inline pytorch_lightning/trainer/optimizers.py:122]: was the fix to rearrange the if-else tower here? Or was anything else needed to change? Hard to see ^^

**Review by awaelchli — APPROVED** [review]

File: docs/source/common/optimizers.rst (hunk 10 above)
Comment (kaushikb11) [reply docs/source/common/optimizers.rst:244]: Updated!

File: tests/trainer/optimization/test_manual_optimization.py (hunk 9 above)
Comment (kaushikb11) [reply tests/trainer/optimization/test_manual_optimization.py]:
```python
def _validate_scheduler_optimizer(optimizers, lr_schedulers):
    if any(sch["scheduler"].optimizer not in optimizers for sch in lr_schedulers):
        raise MisconfigurationException(
            "Some schedulers are attatched with an optimizer that wasn't returned from `configure_optimizers`."
        )
```
It throws an error here.

File: docs/source/common/optimizers.rst [hunk 13]
```diff
@@ -230,6 +230,21 @@ If you want to call ``lr_scheduler.step()`` every ``n`` steps/epochs, do the fol
         if self.trainer.is_last_batch and (self.trainer.current_epoch + 1) % n == 0:
             sch.step()
 
+If you want to call schedulers that require a metric value after each epoch, consider doing the following:
+
+.. testcode::
+
+    def __init__(self):
```
Comment (Borda) [inline docs/source/common/optimizers.rst:237]: can we add class too?

File: pytorch_lightning/trainer/optimizers.py [hunk 14]
```diff
@@ -144,30 +131,49 @@ def configure_schedulers(
...
+                    extra_keys = [k for k in scheduler.keys() if k not in default_config.keys()]
+                    if extra_keys:
+                        rank_zero_warn(f"Found unsupported keys in the lr scheduler dict: {extra_keys}", RuntimeWarning)
+                    if "scheduler" not in scheduler:
+                        raise MisconfigurationException(
+                            'The lr scheduler dict must have the key "scheduler" with its item being an lr scheduler'
+                        )
+                    if "interval" in scheduler and scheduler["interval"] not in ("step", "epoch"):
+                        raise MisconfigurationException(
+                            f'The "interval" key in lr scheduler dict must be "step" or "epoch"'
```
Comment (Borda) [inline pytorch_lightning/trainer/optimizers.py:150]:
```suggestion
                            'The "interval" key in lr scheduler dict must be "step" or "epoch"'
```

**Review by Borda — COMMENTED** [review]
do we have tests for verification that all the raised exceptions are correct?

File: pytorch_lightning/trainer/optimizers.py (hunk 12 above)
Comment (kaushikb11) [reply pytorch_lightning/trainer/optimizers.py:122]: Nothing has been changed. The exceptions are just being omitted for `manual_optimiation`

File: docs/source/common/optimizers.rst (hunk 13 above)
Comment (kaushikb11) [reply docs/source/common/optimizers.rst:237]: I don't think it's required.

File: pytorch_lightning/trainer/optimizers.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/optimizers.py:122]: okay thanks for confirming, I wasn't sure after 10 mins looking at it ^^

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/trainer/optimization/test_manual_optimization.py (hunk 11 above)
Comment (kaushikb11) [reply tests/trainer/optimization/test_manual_optimization.py:977]: Not really. Updated!

**Review by SeanNaren — APPROVED** [review]

File: docs/source/common/optimizers.rst (hunk 13 above)
Comment (kaushikb11) [reply docs/source/common/optimizers.rst:237]: Following the consistency with other demo codes in the doc.
